File: solutions/javascript/class-design/implementQueueUsingStacks.js
const peek = (stack) => {
    if (stack.length === 0) throw new Error('cannot peek, stack is empty');

    const lastIdx = stack.length - 1;
    return stack[lastIdx];
};

class MyQueue {
    constructor() {
        this.size = 0;
        this.stack1 = [];
        this.stack2 = [];
    }

    push(element) {
        this.stack1.push(element);
        this.size++;
    }

    moveToStack2() {
        while (this.stack1.length > 0) {
            const newElement = this.stack1.pop();
            this.stack2.push(newElement);
        }
    }

    pop() {
        if (this.empty()) throw new Error('cannot pop, queue is empty');

        if (this.stack2.length === 0) {
            this.moveToStack2();
        }

        this.size--;
        return this.stack2.pop();
    }

    peek() {
        if (this.empty()) throw new Error('cannot peek, queue is empty');

        if (this.stack2.length === 0) {
            this.moveToStack2();
        }

        return peek(this.stack2);
    }

    empty() {
        return this.size === 0;
    }
}

File: solutions/python/class-design/implement_queue_using_stacks.py
class MyQueue:
    def __init__(self):
        self.size = 0
        self.stack1 = []
        self.stack2 = []

    def push(self, element):
        self.stack1.append(element)
        self.size += 1

    def moveToStack2(self):
        while len(self.stack1) > 0:
            newElement = self.stack1.pop()
            self.stack2.append(newElement)

    def pop(self):
        if self.empty():
            raise Exception('cannot pop, queue is empty')

        if len(self.stack2) == 0:
            self.moveToStack2()

        self.size -= 1
        return self.stack2.pop()

    def peek(self):
        if self.empty():
            raise Exception('cannot peek, queue is empty')

        if len(self.stack2) == 0:
            self.moveToStack2()

        return self.stack2[-1]

    def empty(self):
        return self.size == 0
